Full Ableton Link support

JoergAtGithub opened the pull request:
Based on the old PR #3783 from @spensbot, I created a new PR with Ableton Link support for Main.
![grafik](https://user-images.githubusercontent.com/64457745/197616811-505bc217-a004-4701-b926-674f2924e0fe.png)
The LINK-Button enables/disables the Link session. The number right to it shows the number of connected Link peers.
It syncronizes Tempo and Beat-Position (Phase).

In Ableton Link all peers are equal, there is no leader. Therefore syncronization is always bidirectional.

In Mixxx, it syncs the Sync-Leader (the deck with SYNC and Crown-Button active) with the other peers.

The most important part of this PR is the hosttime filter implemetation, which provides a jitter free timestamp for each audio buffer start - this timestamp is in host time (computer clock not soundcard clock) which is the needed input for the Ableton Link library.

What does not always work automatical, is to determine the latency of the Sound IO. This is because portaudio does not report reliable latency information for all sound apis. On my system it works well for ASIO API, but not so well for the others. Therefore I added a manual latency compensation setting in the preferences, to allow the user to correct the latency:
<img width="1014" height="446" alt="grafik" src="https://github.com/user-attachments/assets/38079198-4308-44ef-b413-18e3ae1ead50" />


I tested this with the official LinkHut-App of the Ableton Link package and with the Android App "G-Stomper Rythm", which is a free Drummachine, but with advertizing. A very cool Ableton Link software is also the music visualisation and lighting software from spensbot: https://captivatesynth.com/

**Test Plan**
Any Ableton Link software, should pass the following test plan:
https://github.com/Ableton/link/blob/master/TEST-PLAN.md

To install the official Linkhut reference tools used in the test plan, you can use VCPKG:  `./vcpkg install ableton[hut,hutsilent]`

Closes #8650

obsoleszenz commented:
Awesome! <3

octopuscabbage commented:
Please finish merging this! I'm probably going to try bulding this branch tomorrow. I would use ableton link a lot.

daschuer commented on src/engine/sync/abletonlink.cpp:
```diff
@@ -0,0 +1,248 @@
+    m_pLink = std::make_unique<ableton::Link>(120.);
+    m_timeAtStartCallback = m_pLink->clock().micros();
+
+    m_pLinkButton = new ControlPushButton(ConfigKey(group, "sync_enabled"));
```
Please use std::make_unique here as well.

daschuer commented on src/engine/sync/abletonlink.cpp:
```diff
@@ -0,0 +1,248 @@
+            &ControlObject::valueChanged,
+            this,
+            &AbletonLink::slotControlSyncEnabled,
+            Qt::DirectConnection);
```
Why is here a direct connection? This will call the slot form the GUI and the controller thread which may cause a race condition.

daschuer commented on src/engine/sync/abletonlink.cpp:
```diff
@@ -0,0 +1,248 @@
+                        .count();
+}
+
+void AbletonLink::nonAudioPrint() {
```
The function name does not reveal what the function does. Please rename.

daschuer commented on src/engine/sync/abletonlink.h:
```diff
@@ -0,0 +1,137 @@
+    // Link getters to call from audio thread.
+    void audioSafePrint();
+
+    /// Link getters to call from non-audio thread.
```
I do not understand this comment and the comment above. Please improve.

daschuer commented on src/engine/sync/abletonlink.cpp:
```diff
@@ -0,0 +1,248 @@
+
+// ***************************
+// Debug output functions only
+// ***************************
```
We don't use ascii decorations in Mixxx. Please remove.

daschuer commented on src/engine/sync/abletonlink.cpp:
```diff
@@ -0,0 +1,248 @@
+// Debug output functions only
+// ***************************
+
+void AbletonLink::audioSafePrint() {
```
The function name should also be improved.

daschuer commented on src/engine/sync/abletonlink.cpp:
```diff
@@ -0,0 +1,248 @@
+            (m_pEngineSync->getLeaderChannel()->getEngineBuffer() != nullptr)) {
+        // Only set a new buffer size if we can determine it, otherwise we assume that it's not changed
+        m_currentLatency =
+                std::chrono::microseconds(m_pEngineSync->getLeaderChannel()
```
Here is would be better to cache the checked pointers in a local variable. This way it makes sure that the value is not changed after the null check.

daschuer commented on src/engine/sync/abletonlink.h:
```diff
@@ -0,0 +1,137 @@
+
+    mixxx::Bpm m_oldTempo;
+
+    std::chrono::microseconds m_currentLatency;
```
Latency is ambiguous. In this case it is the buffer size. Please rename.

daschuer commented on src/engine/sync/abletonlink.cpp:
```diff
@@ -0,0 +1,248 @@
+// Approximate the system time when the first sample in the current audio buffer will hit the speakers
+std::chrono::microseconds AbletonLink::getHostTimeAtSpeaker(
+        const std::chrono::microseconds hostTime) const {
+    return hostTime - m_currentLatency;
```
I think we need to use here the time when the samples reach the DAC. 
See: VisualPlayPosition::setCallbackEntryToDacSecs

daschuer commented on src/engine/sync/abletonlink.cpp:
```diff
@@ -0,0 +1,248 @@
+}
+
+void AbletonLink::onCallbackStart(int sampleRate, int bufferSize) {
+    /// Uses ableton's HostTimeFilter class to create a smooth linear regression between absolute sample time and system time
```
Please break this long line.

daschuer reviewed (changes requested):
Thank you for this nice PR
Here some comments:

Holzhaus commented on src/engine/enginemaster.cpp:
```diff
@@ -268,9 +268,11 @@ const CSAMPLE* EngineMaster::getSidechainBuffer() const {
 
-void EngineMaster::processChannels(int iBufferSize) {
+void EngineMaster::processChannels(int iBufferSize,
+        std::chrono::microseconds absTimeWhenPrevOutputBufferReachsDac) {
```
```suggestion
        std::chrono::microseconds absTimeWhenPrevOutputBufferReachesDac) {
```
?

fwcd commented:
Any updates on this? Super excited by Link support in Mixxx

JoergAtGithub commented:
Try the builds of this PR - it should work in general.

fwcd commented:
Pretty cool, I've tested it and tempo sync seems to work pretty well! Two things:

- Mixxx seems to freeze occasionally during a session, I've tested it with two Mixxx clients talking to each other, but it seems to happen with other Ableton Link clients too
- I haven't gotten phase sync to work, neither with nor without quantization enabled (I'm testing 2 Mixxx instances, on separate machines, each with a single track playing and sync crown enabled). Perhaps I'm missing something here?

Also I think this would be a prime candidate for bar detection if/when Mixxx finally gets it (#4489).

JoergAtGithub commented:
I updated this PR to my local development state.
If phase doesn't sync, try to scratch the waveform by mouse, does the phase sync then?

varoudis commented:
Any Mac Builds I can test? This is soo cool!

fwcd commented:
> Any Mac Builds I can test? This is soo cool!

Check the GitHub Actions builds: https://github.com/mixxxdj/mixxx/actions/runs/4558513041

luzpaz commented:
Is there an ETA on this ?

JoergAtGithub commented:
You can download a build of this PR now. General availability depends much on testing on different platforms with different sound driver configurations. It's developed under Windows11 with ASIO sounddriver. I would be happy if users of other OSes would execute the Ableton Link test plan: https://github.com/Ableton/link/blob/master/TEST-PLAN.md

hellocatfood commented:
Where can this build be downloaded from?

fwcd commented:
@hellocatfood https://github.com/mixxxdj/mixxx/actions/runs/5148635078?pr=10999

ronso0 commented:
> > Any Mac Builds I can test? This is soo cool!
> 
> Check the GitHub Actions builds: https://github.com/mixxxdj/mixxx/actions/runs/4558513041

and scroll down to Artifacts. https://github.com/mixxxdj/mixxx/wiki/Testing#github-pull-requests
As with any development branch, make sure to back up your database and settings (see first section in that wiki page).

fwcd commented:
The old build artifacts have unfortunately expired by now, so it needs to be recent-ish.

ronso0 commented:
@JoergAtGithub could you merge main to trigger builds?

ronso0 commented:
Whoopsy, either the merge went wrong somehow or main is broken.

JoergAtGithub commented:
It's just an added check for automock header, which this PRs code doesn't pass.

JoergAtGithub commented:
New test builds are available for download now!

hellocatfood commented:
Will there be any test builds for Ubuntu 23.04?

Swiftb0y commented on CMakeLists.txt:
```diff
@@ -2030,6 +2032,27 @@ if(WIN32)
+  message(STATUS "Using existing system installation of Ableton Link")
+  target_link_libraries(mixxx-lib PRIVATE Ableton::Link)
+else()
+  message(STATUS "Using internal Ableton Link")
+  include(${CMAKE_CURRENT_SOURCE_DIR}/lib/ableton-link/share/ableton/AbletonLinkConfig.cmake)
+  target_link_libraries(mixxx-lib PUBLIC Ableton::Link)
+  target_include_directories(mixxx-lib SYSTEM PUBLIC lib/ableton-link/include/)
+endif()
```
fyi this doesn't work. Did you forget to include the `AbletonLinkConfig.cmake` or whats going on here? Do we even want to fendor this? availability is still quite sparse on linux unfortunately. https://repology.org/project/ableton-link/versions

JoergAtGithub replied:
I don't understand the point. It builds on all three  platforms including Ubuntu. If the package is not available on unsuported third party Linux distributions, it's the job of these distributions maintainers to create such a package.

Swiftb0y replied:
Okay thanks for clarifying. I was under the impression that we would fallback to a vendored version of ableton-link if it was not available on the system. Similar to how we do it with libshout-idjc or portaudio or any other library in our `lib/` folder. That usually works, except in this case the files for that are obviously missing from the lib folder and thus CMake just plain fails, which is definitely a bug. So either we fix the internal fallback, or remove the feature entirely.

JoergAtGithub commented on CMakeLists.txt:
```diff
@@ -1205,7 +1206,8 @@ if(MSVC)
   else()
     target_compile_options(mixxx-lib PUBLIC /W3)
-    target_compile_definitions(mixxx-lib PUBLIC _SILENCE_CXX17_ITERATOR_BASE_CLASS_DEPRECATION_WARNING _CRT_SECURE_NO_WARNINGS)
+    # _HAS_DEPRECATED_RESULT_OF is needed to compile AbletonLink with MSVC in C++20 mode (upstream fix is expected in AbletonLink 3.0.6: https://github.com/Ableton/link/issues/119)
+    target_compile_definitions(mixxx-lib PUBLIC _SILENCE_CXX17_ITERATOR_BASE_CLASS_DEPRECATION_WARNING _HAS_DEPRECATED_RESULT_OF _CRT_SECURE_NO_WARNINGS)
```
It depends on which Ubuntu version we do need to support when we merge this. For now we need it.

Swiftb0y commented:
I would like to test this but in order to do so easily and in order for this PR to be merged we'd the ableton-link library needs to be more available (such as on fedora). I see two options: either we vendor it as we do with the other dependencies in `lib/` or we do it properly and package it for other distributions. @uklotzde would you be able to take care of packaging it for fedora?

JoergAtGithub commented:
@Swiftb0y I'm currently looking if I can use your new timer implementation, as reference for Ableton Link too. They recommend, to use the same clock accross the application - currently this PR uses, what is provided by Ableton as reference implementation.

Swiftb0y commented:
Not sure if that makes much sense... You could just use use `ableton::platform::stl::Clock` or a generic reimplementation thereof that uses `PerformanceTimer::ClockT` as its underyling clock if you want to ensure that `PerformanceTimer` and the abletonclock use the same source. Or you could add another member function to PerformanceTimer `std::chrono::microseconds micros() const {return std::duration_cast<std::chrono::microseconds>(ClockT::now())};` but that does not really fit the rest of the interface and is somewhat contrived and intrusive. I'd prefer the reimplementation that depends on `PerformanceTimer::ClockT`.

uklotzde commented:
> @uklotzde would you be able to take care of packaging it for fedora?

Neither time nor interest to do this. Since it is a header-only library vendoring is the right choice to ensure compatibility across all platforms. That's also how I understand their advice in the README, see https://github.com/Ableton/link#building-link.

Swiftb0y commented:
Thank you for your insight, then vendoring it is then I guess. I may look into it.

JoergAtGithub commented:
The predeccessor #3783 of this PR of this had a vendored copy. Removing this vendored copy was requested in that PR, because it contained hundreds of additional files.
Better add a download fallback, as we have it for libdjinterop and libkeyfinder.

JoergAtGithub commented:
@Swiftb0y Could you try if the latest commit builds on Fedora Linux

Swiftb0y commented:
Pretty sure it builds (currently at 50%) but it also builds all the ableton link tests and linkhut, etc. It also prints a bunch of warnings (pretty sure not our fault, see below). The biggest issue I see is that FetchContent is not compatible with offline build infrastructures such as koji (fedora). Instead we'd need to use the slightly more verbose `ExternalProject_Add` (see the keyfinder section in our CMakeLists.txt for an example). 

<details><summary>Cmake Warnings</summary>
<p>

```
CMake Warning (dev) at cbuild/_deps/abletonlink-src/extensions/abl_link/abl_link.cmake:5 (add_library):
  Policy CMP0063 is not set: Honor visibility properties for all target
  types.  Run "cmake --help-policy CMP0063" for policy details.  Use the
  cmake_policy command to set the policy and suppress this warning.

  Target "abl_link" of type "STATIC_LIBRARY" has the following visibility
  properties set for CXX:

    CXX_VISIBILITY_PRESET
    VISIBILITY_INLINES_HIDDEN

  For compatibility CMake is not honoring them for this target.
Call Stack (most recent call first):
  cbuild/_deps/abletonlink-src/CMakeLists.txt:36 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at cbuild/_deps/abletonlink-src/src/CMakeLists.txt:86 (add_executable):
  Policy CMP0063 is not set: Honor visibility properties for all target
  types.  Run "cmake --help-policy CMP0063" for policy details.  Use the
  cmake_policy command to set the policy and suppress this warning.

  Target "LinkCoreTest" of type "EXECUTABLE" has the following visibility
  properties set for CXX:

    CXX_VISIBILITY_PRESET
    VISIBILITY_INLINES_HIDDEN

  For compatibility CMake is not honoring them for this target.
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at cbuild/_deps/abletonlink-src/src/CMakeLists.txt:102 (add_executable):
  Policy CMP0063 is not set: Honor visibility properties for all target
  types.  Run "cmake --help-policy CMP0063" for policy details.  Use the
  cmake_policy command to set the policy and suppress this warning.

  Target "LinkDiscoveryTest" of type "EXECUTABLE" has the following
  visibility properties set for CXX:

    CXX_VISIBILITY_PRESET
    VISIBILITY_INLINES_HIDDEN

  For compatibility CMake is not honoring them for this target.
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at cbuild/_deps/abletonlink-src/examples/CMakeLists.txt:182 (add_executable):
  Policy CMP0063 is not set: Honor visibility properties for all target
  types.  Run "cmake --help-policy CMP0063" for policy details.  Use the
  cmake_policy command to set the policy and suppress this warning.

  Target "LinkHut" of type "EXECUTABLE" has the following visibility
  properties set for CXX:

    CXX_VISIBILITY_PRESET
    VISIBILITY_INLINES_HIDDEN

  For compatibility CMake is not honoring them for this target.
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at cbuild/_deps/abletonlink-src/examples/CMakeLists.txt:209 (add_executable):
  Policy CMP0063 is not set: Honor visibility properties for all target
  types.  Run "cmake --help-policy CMP0063" for policy details.  Use the
  cmake_policy command to set the policy and suppress this warning.

  Target "LinkHutSilent" of type "EXECUTABLE" has the following visibility
  properties set for CXX:

    CXX_VISIBILITY_PRESET
    VISIBILITY_INLINES_HIDDEN

  For compatibility CMake is not honoring them for this target.
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at cbuild/_deps/abletonlink-src/extensions/abl_link/CMakeLists.txt:14 (add_executable):
  Policy CMP0063 is not set: Honor visibility properties for all target
  types.  Run "cmake --help-policy CMP0063" for policy details.  Use the
  cmake_policy command to set the policy and suppress this warning.

  Target "link_hut" of type "EXECUTABLE" has the following visibility
  properties set for C:

    C_VISIBILITY_PRESET

  For compatibility CMake is not honoring them for this target.
This warning is for project developers.  Use -Wno-dev to suppress it.
```

</p>
</details>

Swiftb0y commented:
compiles successfully. Starting two mixxx instances and enabling link successfully connects them to each other but there are a couple of bugs (though I'm not sure if all of them related to Qt6 or this PR). One thing I can reproduce pretty clearly is that when I start two instances, activate link, load a single track in each and then activate sync in one instance, it syncs to 120bpm instead of clock of the other instance. I need to play around a little, loading other tracks and so forth until I can get it to work properly.

JoergAtGithub commented:
Did you enable the explicit sync leader (crown button)?

Swiftb0y commented:
I don't think so. tbh I have not experimented much with explicit leader, so I don't even know what the expected behavior should be.

JoergAtGithub commented:
This PR currently syncs the explicit leader - without nothing will happen.

Swiftb0y commented:
Thanks for clarifying, I'll try to understand how its supposed to work properly from the implementation and retest soon.

luzpaz commented:
Soft bump

fwcd commented:
Any blockers here? Would be thrilled about Ableton Link support, let me know if there's anything to review or test

JoergAtGithub commented:
I need to rework this, since it is based on the Explicit Sync Leader functionality, which was removed short before the 2.4 release.

han-so1omon commented:
@JoergAtGithub What is the current state of this? I may be interested to pick it up, depending on the time commitments required

JoergAtGithub commented:
If you could test it and measure the delay with different SoundAPIs especially on macOS and Linux would help. For me it works quite well with an ASIO driver  on Windows 11.
Especially do all the required tests according https://github.com/Ableton/link/blob/master/TEST-PLAN.md would be of help.

han-so1omon commented:
@JoergAtGithub I'm getting some issues with the initial build step on MacOS 14. https://github.com/mixxxdj/mixxx/issues/13387

luzpaz commented:
Any progress here ?

han-so1omon commented:
I almost picked it up, but I was having issues compiling, so I dropped it.
If I can get support on that I might be able to pick it up

El jue, 18 jul 2024, 2:44 p. m., luzpaz ***@***.***> escribió:

> Any progress here ?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/mixxxdj/mixxx/pull/10999#issuecomment-2237673531>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABUKJ6JXEJO5YLZHXFIBQN3ZNAZKHAVCNFSM6AAAAAARNKDRP2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDEMZXGY3TGNJTGE>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

han-so1omon commented:
@fwcd also very interested. It seems to be nearly complete. I tried to do testing but couldn't get compile cleaning on Mac. Any interest in collaborating?

Swiftb0y commented on src/soundio/sounddeviceportaudio.cpp:
```diff
@@ -94,7 +97,9 @@ SoundDevicePortAudio::SoundDevicePortAudio(UserSettingsPointer config,
-          m_lastCallbackEntrytoDacSecs(0) {
+          m_lastCallbackEntrytoDacSecs(0),
+          m_cummulatedBufferTime(0),
+          m_meanOutputLatency(MovingInterquartileMean(501)) {
```
wheres the 501 coming from?

Swiftb0y commented on src/engine/sync/abletonlink.h:
```diff
@@ -0,0 +1,142 @@
+    std::unique_ptr<ableton::Link> m_pLink;
+    ableton::link::HostTimeFilter<ableton::link::platform::Clock> m_hostTimeFilter;
+    const QString m_group;
+    EngineSync* m_pEngineSync;
```
```suggestion
    EngineSync* m_pEngineSync; // unowned, must outlive this.
```

Swiftb0y commented on src/engine/sync/abletonlink.h:
```diff
@@ -0,0 +1,142 @@
+
+    // Test/Debug code
+
+    QTimer* m_pTestTimer;
```
```suggestion
    std::parented_ptr<QTimer> m_pTestTimer;
```

Swiftb0y reviewed (commented):
couple comments, to I really am not the right person to review this because I have zero knowledge EngineSync and Syncable code...

fwcd commented on src/engine/sync/abletonlink.cpp:
```diff
@@ -0,0 +1,232 @@
+}
+
+void AbletonLink::slotControlSyncEnabled(double value) {
+    qDebug() << "AbletonLink::slotControlSyncEnabled" << value;
```
Sorry if I'm missing something, but isn't that exactly the point of it? The debug messages shouldn't show up in production logs and thus not add much overhead in that case.

Swiftb0y replied:
yes, debug logs are general a good thing. But this line in particular does make much sense because it seems like a leftover from printf-debugging:
1. new code made for properly debugging user issues should leverage logging categories. Logging categories are overkill for simple printf-debugging but are really useful in logs IMO.
2. This line is just logging whether a slot is called. this doesn't seem particularly useful in the general case where we want debug messages to show up in user logs so we can trace an issue.

fwcd commented:
I have resolved the merge conflict in `CMakeLists.txt` (in case anyone wants to review that: `git show --remerge-diff fb6173bf50` will show only the conflict resolution). There are still some style checks failing though which I didn't tackle.

JoergAtGithub commented:
I intentionally did not merged Main ,because the "Explicit Sync Leader" functionality, where this functionality is based on, was removed temporary before the 2.4 release. Now it can't work anymore.

fwcd commented:
Oh sorry for that, I'll force-push the previous state back

JoergAtGithub commented:
Thans, I will look for an update myself. But what would really help, if you could measure the absolute time difference between a second device and the Mac running Mixxx. The specification say, that it must not be more than 5ms. I achieved this under windows with ASIO sounddriver API, but not with other sounddrivers.
If anyone on macOS could test sample the audio ouptput of Mixxx together with the output of the official reference application LinkHut (ideally running on a second device). This would be a valuable information.

fwcd commented:
How do I measure that? I couldn't find anything in the macOS GUI for LinkHut, does the command-line tool output more the latencies?

JoergAtGithub commented:
LinkHut is a command line tool, that come with AbletonLink. I added it also as feature to the VCPKG port. It just generated some beeps. If you store this similar beeps as soundfile and set the beatgrid accordingling, the two sound outputs should be in sync. You can sample the outputs of both devices with two input channels of any audio interface.

JoergAtGithub commented on src/engine/sync/enginesync.h:
In the engine thread

fwcd commented on src/engine/sync/abletonlink.h:
```diff
@@ -0,0 +1,135 @@
+// Note that the resolution of std::chrono::steady_clock is not guaranteed
+// to be high resolution, but it is guaranteed to be monotonic.
+// However, on all major platforms, it is high resolution enough.
+using AbletonLinkClockReference = ableton::platforms::stl::Clock;
```
I think you could also use `ableton::Link::Clock` which would use a platform-specific high resolution monotonic clock (on Windows and macOS IIRC) and only fall back to the STL clock if unavailable. Or do we specifically need the STL clock?

JoergAtGithub replied:
We use this already at other places on Mixxx, and I want to have these filtered absolute time stamps to be useable anywhere (e.g in a future VSync PR). This is why I changed this.

fwcd replied:
Makes sense!

fwcd commented on src/engine/sync/abletonlink.h:
```diff
@@ -0,0 +1,135 @@
+
+#include <ableton/Link.hpp>
+#include <ableton/link/HostTimeFilter.hpp>
+#include <ableton/platforms/stl/Clock.hpp>
```
`ableton::Link::Clock` should also be included in `Link.hpp`, not sure if these platform-specific headers are intended to be consumed externally (since the whole library is header-only)

JoergAtGithub replied:
No, does not work. This was the first I tried.

jonahclarsen commented:
Just want to say I'm super stoked for this and thankful for all your effort!

Swiftb0y commented on src/engine/sync/abletonlink.cpp:
```diff
@@ -0,0 +1,217 @@
+            &AbletonLink::slotControlSyncEnabled,
+            Qt::DirectConnection);
+
+    m_pNumLinkPeers->setReadOnly();
+    m_pNumLinkPeers->forceSet(0);
+
+    // The callback is invoked on a Link - managed thread.
+    // The callback is the only entity, which access m_pNumLinkPeers
+    m_pLink->setNumPeersCallback([this](std::size_t numPeers) {
+        m_pNumLinkPeers->forceSet(numPeers);
+    });
```
then make it a member of the lambda instead.
```suggestion
    // The callback is invoked on a Link - managed thread.
    m_pLink->setNumPeersCallback([ControlObject numLinkPeers = []{
        ControlObject numLinkPeers(ConfigKey(group, numPeers));
        numLinkPeers.setReadOnly();
        numLinkPeers.forceSet(0);
        return numLinkPeers;
    }()](std::size_t numPeers) {
        numLinkPeers.forceSet(numPeers);
    });
```

JoergAtGithub replied:
The proposed code does not compile, as it does not match to the declaration in the headers of Link.

Swiftb0y replied:
that does not make sense. Can you post the error?

JoergAtGithub replied:
![grafik](https://github.com/user-attachments/assets/421e4425-13aa-4c68-819c-7c47a3a41536)

Swiftb0y replied:
Thanks. looks like a syntax issue. I'll figure it out.

Swiftb0y replied:
init captures don't support explicit type declarations. Try this:
```suggestion
    // The callback is invoked on a Link - managed thread.
    m_pLink->setNumPeersCallback([numLinkPeers = []{
        ControlObject numLinkPeers(ConfigKey(group, numPeers));
        numLinkPeers.setReadOnly();
        numLinkPeers.forceSet(0);
        return numLinkPeers;
    }()](std::size_t numPeers) {
        numLinkPeers.forceSet(numPeers);
    });
```

JoergAtGithub replied:
This doesn't work too.

Swiftb0y replied:
```suggestion
    // The callback is invoked on a Link - managed thread.
    m_pLink->setNumPeersCallback([numLinkPeers = [&]{
        ControlObject numLinkPeers(ConfigKey(group, numPeers));
        numLinkPeers.setReadOnly();
        numLinkPeers.forceSet(0);
        return numLinkPeers;
    }()](std::size_t numPeers) mutable {
        numLinkPeers.forceSet(numPeers);
    });
```

JoergAtGithub replied:
Doesn't work:
```
  FAILED: CMakeFiles/mixxx-lib.dir/src/engine/sync/abletonlink.cpp.obj 
  C:\PROGRA~1\MICROS~1\2022\COMMUN~1\VC\Tools\MSVC\1441~1.341\bin\Hostx64\x64\cl.exe  /nologo /TP -DAMD64 -DDJINTEROP_SOURCE -DLINK_PLATFORM_WINDOWS=1 -DMIXXX_BUILD_RELEASE -DMIXXX_DEBUG_ASSERTIONS_ENABLED -DMIXXX_DEBUG_ASSERTIONS_FATAL -DMIXXX_USE_QML -DMIXXX_USE_QOPENGL -DNOMINMAX -DPROTOBUF_USE_DLLS -DQT_CONCURRENT_LIB -DQT_CORE5COMPAT_LIB -DQT_CORE_LIB -DQT_GUI_LIB -DQT_LABSQMLMODELS_LIB -DQT_NETWORK_LIB -DQT_NO_DEBUG -DQT_OPENGL_LIB -DQT_PRINTSUPPORT_LIB -DQT_QMLINTEGRATION_LIB -DQT_QMLMODELS_LIB -DQT_QMLWORKERSCRIPT_LIB -DQT_QML_LIB -DQT_QUICKCONTROLS2IMPL_LIB -DQT_QUICKCONTROLS2_LIB -DQT_QUICKLAYOUTS_LIB -DQT_QUICKSHAPES_LIB -DQT_QUICKTEMPLATES2_LIB -DQT_QUICKWIDGETS_LIB -DQT_QUICK_LIB -DQT_SHADERTOOLS_LIB -DQT_SQL_LIB -DQT_SVGWIDGETS_LIB -DQT_SVG_LIB -DQT_TABLET_SUPPORT -DQT_TESTCASE_BUILDDIR=\"D:/mixxx/build/x64__portable\" -DQT_TESTCASE_SOURCEDIR=\"D:/mixxx\" -DQT_TESTLIB_LIB -DQT_USE_QSTRINGBUILDER -DQT_WIDGETS_LIB -DQT_XML_LIB -DSFC_SUPPORTS_SET_COMPRESSION_LEVEL -DUNICODE -DWIN32 -DWIN32_LEAN_AND_MEAN -DWIN64 -DWINVER=0x0601 -D_ATL_MIN_CRT -D_CRT_SECURE_NO_WARNINGS -D_ENABLE_EXTENDED_ALIGNED_STORAGE -D_SILENCE_CXX17_ITERATOR_BASE_CLASS_DEPRECATION_WARNING -D_UNICODE -D_USE_MATH_DEFINES -D_WIN32_WINNT=0x0601 -D_WIN64 -D__BATTERY__ -D__BROADCAST__ -D__BULK__ -D__ENGINEPRIME__ -D__FFMPEG__ -D__HID__ -D__HSS1394__ -D__LILV__ -D__MAD__ -D__MEDIAFOUNDATION__ -D__MODPLUG__ -D__OPUS__ -D__PORTMIDI__ -D__QTKEYCHAIN__ -D__RUBBERBAND__ -D__SNDFILE__ -D__SQLITE3__ -D__SSE2__ -D__SSE__ -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -D__STEM__ -D__VINYLCONTROL__ -D__WINDOWS__ -D__WV__ -ID:\mixxx\build\x64__portable\mixxx-lib_autogen\include -ID:\mixxx\src -ID:\mixxx\build\x64__portable\src -external:ID:\mixxx\lib\fidlib -external:ID:\mixxx\lib\portaudio -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include -external:ID:\mixxx\lib\rigtorp\SPSCQueue\include -external:ID:\mixxx\lib\replaygain -external:ID:\mixxx\lib\reverb -external:ID:\mixxx\lib\libshout-idjc\include -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\share\ableton-link\..\..\include\ableton -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\share\asio\..\..\include -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\djinterop -external:ID:\mixxx\lib\rekordbox-metadata -external:ID:\mixxx\lib\kaitai -external:ID:\mixxx\lib\mp3guessenc-0.27.4 -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtConcurrent -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6 -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtCore -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\share\Qt6\mkspecs\win32-msvc -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtGui -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtNetwork -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtOpenGL -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtPrintSupport -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtWidgets -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtQml -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtQmlIntegration -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtSql -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtSvg -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtTest -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtXml -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtShaderTools -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtSvgWidgets -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtCore5Compat -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtQuick -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtQmlModels -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtLabsQmlModels -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtQmlModels\6.5.3 -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtQmlModels\6.5.3\QtQmlModels -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtCore\6.5.3 -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtCore\6.5.3\QtCore -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtQml\6.5.3 -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtQml\6.5.3\QtQml -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtQuickControls2 -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtQuickControls2Impl -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtQuickLayouts -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtGui\6.5.3 -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtGui\6.5.3\QtGui -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtQuick\6.5.3 -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtQuick\6.5.3\QtQuick -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtQuickShapes -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtQuickShapes\6.5.3 -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtQuickShapes\6.5.3\QtQuickShapes -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtQuickTemplates2 -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtQuickWidgets -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtQmlWorkerScript -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\taglib -external:ID:\mixxx\lib\qm-dsp -external:ID:\mixxx\lib\qm-dsp\include -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\opus -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\hidapi -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\libusb-1.0 -external:ID:\mixxx\lib\xwax -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\wavpack -external:W0 /DWIN32 /D_WINDOWS /EHsc /Zc:__cplusplus /Zc:inline /Z7 /O2  /DNDEBUG -std:c++20 -MD /UTF8 /fp:fast /Gy /W3 /wd4200 -DPA_USE_JACK=1 -DPA_USE_ASIO=1 -DPA_USE_DS=1 -DPA_USE_WMME=1 -DPA_USE_WASAPI=1 -DPA_USE_WDMKS=1 -Zc:__cplusplus -permissive- -utf-8 /YuD:/mixxx/build/x64__portable/CMakeFiles/mixxx-lib.dir/cmake_pch.hxx /FpD:/mixxx/build/x64__portable/CMakeFiles/mixxx-lib.dir/./cmake_pch.cxx.pch /FID:/mixxx/build/x64__portable/CMakeFiles/mixxx-lib.dir/cmake_pch.hxx /showIncludes /FoCMakeFiles\mixxx-lib.dir\src\engine\sync\abletonlink.cpp.obj /Fdmixxx-lib.dir\mixxx-lib.pdb /FS -c D:\mixxx\src\engine\sync\abletonlink.cpp
D:\mixxx\src\engine\sync\abletonlink.cpp(27): error C2614: 'AbletonLink': illegal member initialization: 'm_pNumLinkPeers' is not a base or member
D:\mixxx\src\engine\sync\abletonlink.cpp(47): error C2280: 'ControlObject::ControlObject(ControlObject &&)': attempting to reference a deleted function
  D:\mixxx\src\control/controlobject.h(189): note: see declaration of 'ControlObject::ControlObject'
```
I suggest you compile it yourself first, because this two person trial and error development seems not very efficient to me.

Swiftb0y replied:
Okay, I've investigated and my conclusion is that moving this into the lambda is infeasible because link requires the lambda to be copy-able. This would require making the member of the lambda a `std::shared_ptr<ControlObject>` which has unnecessary overhead. They also don't allow mutable which is another oversight in the link API. (not that that is required because TIL that you can still call mutable member functions of a pointer managed by a `const unique_ptr`, which seems dangerous). 

The only recommendation I would make then is moving the `m_pNumLinkPeers` above `m_pLink` in the class definition so it doesn't get destroyed while `m_pLink` still references it.

Swiftb0y commented on src/engine/sync/abletonlink.cpp:
```diff
@@ -0,0 +1,217 @@
+    // TODO: Check the special case of half/double BPM sync.
+}
+
+void AbletonLink::reinitLeaderParams(double beatDistance, mixxx::Bpm baseBpm, mixxx::Bpm bpm) {
+    Q_UNUSED(baseBpm)
+    updateLeaderBeatDistance(beatDistance);
+    updateLeaderBpm(bpm);
+}
+
+void AbletonLink::updateInstantaneousBpm(mixxx::Bpm bpm) {
+    Q_UNUSED(bpm)
+}
```
[CppCoreGL F.9](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f9-unused-parameters-should-be-unnamed)
```suggestion
void AbletonLink::reinitLeaderParams(double beatDistance, mixxx::Bpm, mixxx::Bpm bpm) {
    updateLeaderBeatDistance(beatDistance);
    updateLeaderBpm(bpm);
}

void AbletonLink::updateInstantaneousBpm(mixxx::Bpm) {
}
```

Swiftb0y commented on src/engine/sync/abletonlink.cpp:
```diff
@@ -0,0 +1,217 @@
+             << m_absTimeWhenPrevOutputBufferReachesDac.count() -
+                    absTimeWhenPrevOutputBufferReachesDac.count();
+                    */
+    m_audioBufferTimeMicros = std::chrono::microseconds(
+            bufferSize / mixxx::audio::ChannelCount::stereo() /
+            sampleRate * 1000000);
```
this seems unused.

Swiftb0y commented on src/engine/sync/abletonlink.h:
```diff
@@ -0,0 +1,137 @@
+
+    mixxx::Bpm m_oldTempo;
+
+    std::chrono::microseconds m_audioBufferTimeMicros;
+    std::chrono::microseconds m_absTimeWhenPrevOutputBufferReachesDac;
+    std::chrono::microseconds m_sampleTimeAtStartCallback;
+    std::chrono::microseconds m_timeAtStartCallback;
```
can be simplified
* `m_audioBufferTimeMicros` is only set, never read
* `m_sampleTimeAtStartCallback` is only initialized, never changed nor read
* `m_timeAtStartCallback` is only set, never read
```suggestion
    std::chrono::microseconds m_absTimeWhenPrevOutputBufferReachesDac;
```

Swiftb0y commented on src/engine/sync/abletonlink.h:
```diff
@@ -0,0 +1,137 @@
+
+  private:
+    ableton::link::HostTimeFilter<MixxxClockRef> m_hostTimeFilter;
+    const QString m_group;
```
const member variables inhibit the possibility to move.
```suggestion
    QString m_group;
```

JoergAtGithub replied:
But this must not move, it should stay constant for the lifetime of the class instance.

Swiftb0y replied:
Sure it can stay constant, but thats an invariant the class itself needs to hold up. If you mark the member constant, the entire class instance can't move anymore. Here's some more background on const member variables: https://www.sandordargo.com/blog/2020/11/11/when-use-const-2-member-variables

JoergAtGithub replied:
This is declared const in https://github.com/mixxxdj/mixxx/blob/8790b5fbf2eb555da740a16ae39cf08215868031/src/engine/sync/syncable.h#L92 I can not change this modifying other parts of the sync code. And I want not expand the scope of the feature PR.

JoergAtGithub replied:
Reading it again, I was wrong, this was just the method, not the member.

Swiftb0y commented on src/engine/sync/abletonlink.cpp:
```diff
@@ -0,0 +1,217 @@
+
+// Debug output function, to be called in audio thread
+void AbletonLink::audioThreadDebugOutput() {
+    qDebug() << "isEnabled()" << m_pLink->isEnabled();
+    qDebug() << "numPeers()" << m_pLink->numPeers();
+
+    const auto sessionState = m_pLink->captureAudioSessionState();
+
+    qDebug() << "sessionState.tempo()" << sessionState.tempo();
+    qDebug() << "sessionState.beatAtTime()"
+             << sessionState.beatAtTime(m_pLink->clock().micros(), getQuantum());
+    qDebug() << "sessionState.phaseAtTime()"
+             << sessionState.phaseAtTime(m_pLink->clock().micros(), getQuantum());
+    qDebug() << "sessionState.timeAtBeat(0)" << sessionState.timeAtBeat(0.0, getQuantum()).count();
+    qDebug() << "sessionState.isPlaying()" << sessionState.isPlaying();
+    qDebug() << "sessionState.timeForIsPlaying()" << sessionState.timeForIsPlaying().count();
+
+    // Est. Delay (micro-seconds) between onCallbackStart() and buffer's first
+    // audio sample reaching speakers
+    qDebug() << "Est. Delay (us)"
+             << (m_absTimeWhenPrevOutputBufferReachesDac -
+                        m_pLink->clock().micros())
+                        .count();
```
we should also make use of the logger here, or do you disagree?

JoergAtGithub commented on CMakeLists.txt:
```diff
@@ -1630,9 +1631,10 @@ if(MSVC)
                 # Note: Even with CMAKE_C_STANDARD = 99 MSVC does not complain about C99 flexible array members
     )
-    target_compile_definitions(mixxx-lib PUBLIC _SILENCE_CXX17_ITERATOR_BASE_CLASS_DEPRECATION_WARNING _CRT_SECURE_NO_WARNINGS)
+    # _HAS_DEPRECATED_RESULT_OF is needed to compile AbletonLink with MSVC in C++20 mode (upstream fix is expected in AbletonLink 3.0.6: https://github.com/Ableton/link/issues/119)
+    target_compile_definitions(mixxx-lib PUBLIC WIN32_LEAN_AND_MEAN _SILENCE_CXX17_ITERATOR_BASE_CLASS_DEPRECATION_WARNING _HAS_DEPRECATED_RESULT_OF _CRT_SECURE_NO_WARNINGS)
```
Not possible, because Ubuntu 22.04 contains AbletonLink 3.0.3

JoergAtGithub replied:
Should be possible now!

EvanBeierschmitt commented:
Hi, all! Sorry, just jumping in, but what needs to happen before this branch is merged?

JoergAtGithub commented:
> Bitwig and Mixxx are not seeing each other.
> 
> I did get link working with Rekordbox, so I suspect this is a Mixxx bug. How can I grab some logs to share?

I only tested this on Windows, but I could imagine that this are some security/firewall rules that need to be set on macOS additional.

Providing a Mixxx log file would definetally help: https://github.com/mixxxdj/mixxx/wiki/Finding%20the%20Mixxx.log%20file

SephReed commented:
> Providing a Mixxx log file would definetally help: https://github.com/mixxxdj/mixxx/wiki/Finding%20the%20Mixxx.log%20file

There's a lot here.  Anything in particular I should be looking for?  

<img width="522" alt="image" src="https://github.com/user-attachments/assets/29562a77-7e59-4692-a280-7634558cb975" />

JoergAtGithub commented:
It's not enabled, does it change the state if you click on the Link button? Do you get any security warnings, when you enable ot the first time?

SephReed commented:
No security warnings or console output when clicking the link button.  I still haven't gotten Mixxx to compile (MacOS arm is still funky support state).  If I could I would go searching for what console output is supposed to happen when clicking the button.  Or if there is none, maybe add some.

What are your insights here?  Should there be console output when clicking the link button?

JoergAtGithub commented:
Could you please provide the full log file, otherwise it's difficult to debug.

quantverse commented:
Because I got Ableton Move myself, I will test this even more thoroughly later.

EvanBeierschmitt commented:
I'm rooting this on! Does anyone have any idea how much longer until this gets merged? I'm spending money to buy DJ software in the meantime due to link compatibility, but I'll switch back to Mixxx as soon as this makes it into the official version!

Eve00000 commented:
> any idea how much longer until this gets merged ... 

Helping the contributor & testing the PR could help. You can build Mixxx yourself with this PR ot download the installer for your OS on the 'Checks' - tab of this PR -> choose builds -> bottom of the page.

EvanBeierschmitt commented:
I'll do that now! Thanks for the tip. I'll let folks know if I run into any issues.

EvanBeierschmitt commented:
Hm. Looks like it's all expired? Sorry, I'm a novice engineer, so still learning the Git ropes. :)

<img width="1415" alt="Screenshot 2025-04-27 at 11 09 01 AM" src="https://github.com/user-attachments/assets/326f3eb6-01aa-4f5f-8ed8-e368b1fb674f" />

Eve00000 commented:
Indeed, sorry, but I think @JoergAtGithub will push a new commit soon to update/rebuild the PR.
As you commented here, you'lll get a notice.
So you try it you'll have to build Mixxx yourself  (with this PR).
How? -> See the Mixxx-Wiki -> getting Started -> [Using Git](https://github.com/mixxxdj/mixxx/wiki/Using%20Git) will guide you setting up the basic.
If you need help, join us on the [Mixxx-zulip](https://mixxx.zulipchat.com/).

quantverse commented:
Small issue with the skin (the button tooltip does not work correctly).

There is this warning in the logfile:

```
warning [Main] Skin parsing failed at skins:LateNight/toolbar.xml:153 <Number>: Invalid <TooltipId> in skin.xml: num_peers | ./src/skin/legacy/legacyskinparser.cpp:2335
```

tuulikauri commented:
Also rooting this on... will look into building mixxx and testing (on Windows, with Ableton and various Ableton Link hardware..)! Cheers yaull.

JoergAtGithub commented:
The Link button just enables Link. You need to set a Sync Leader by the Crown button as well. And a sync in Ableton Link only happens if you change BPM or phase, just enable Link changes nothing.

mtomt commented:
It works now but is possible to use it without syncing the bpm between the players?

JoergAtGithub commented:
One deck must be the Sync-Leader, otherwise it is not clear wich Mixxx deck Link should sync to. But the other decks do not have to have Sync enabled.
In Ableton Link itself, all peers always have equal status. In Ableton Link iself there is no concept like Sync-Leader and Sync-Follower.

JoergAtGithub committed:
Merge remote-tracking branch 'upstream/main' into abletonlink2

JoergAtGithub committed:
Support Android build

JoergAtGithub committed:
Added explicit destructor to release COs and Link in defined order

JoergAtGithub committed:
Remove unused member vars and surrounding code

JoergAtGithub committed:
Added new constant expression kNumPointsInHostTimeFilter in anonymous namespace with a detailed expalantion as comment

JoergAtGithub committed:
Removed unused function arguments

JoergAtGithub commented:
@tuulikauri Please test again if the hung/crash at shutdown still occurs, I hope it is fixed now.

JoergAtGithub committed:
Implemented AbletonLink widgets for Tango

JoergAtGithub committed:
Add AbletonLink widgets to Deere skin

JoergAtGithub committed:
Added minimal LINK widgets to Shade skin

JoergAtGithub committed:
Merge remote-tracking branch 'upstream/main' into abletonlink2

JoergAtGithub committed:
Implement a setting in the sound preferences, to apply a manual sync latency offset

daschuer commented on src/engine/sync/abletonlink.h:
```diff
@@ -0,0 +1,132 @@
+  private:
+    ableton::link::HostTimeFilter<MixxxClockRef> m_hostTimeFilter;
+    QString m_group;
+    EngineSync* m_pEngineSync; // unowned, must outlive this.
```
```suggestion
    EngineSync* m_pEngineSync; // borrowed, must outlive this.
```

daschuer reviewed (commented):
some more comments. Review is in progress.

JoergAtGithub committed:
Apply suggestions from code review

Co-authored-by: Daniel Schürmann <daschuer@mixxx.org>

JoergAtGithub committed:
Merge remote-tracking branch 'upstream/main' into abletonlink2

JoergAtGithub committed:
Fix flatpack builds

JoergAtGithub committed:
Adjusted order of private vars and private methods in header

JoergAtGithub commented on src/engine/sync/abletonlink.cpp:
```diff
@@ -0,0 +1,216 @@
+                &AbletonLink::slotControlSyncEnabled);
+    }
+
+    // Destroy control objects before releasing Link.
```
I would prefer to leave the three comments as they describe the required order of the three steps

daschuer commented on src/engine/sync/abletonlink.cpp:
```diff
@@ -0,0 +1,208 @@
+    m_pLink->enable(false);
+
+    // Destroy control objects before releasing Link.
+    m_pNumLinkPeers.reset();
```
Here m_pNumLinkPeers, becomes null, but m_pLink has still access via the callback. Why is it required that it is destroyed first? It is not the safest options as least.

JoergAtGithub replied:
Indeed a corrected the destruction order and adjusted the comment!

daschuer commented on src/soundio/sounddeviceportaudio.cpp:
```diff
@@ -1142,6 +1166,54 @@ void SoundDevicePortAudio::updateCallbackEntryToDacTime(
+            timeSinceLastCbSecs < bufferSizeSec * 2) {
+        m_meanOutputLatency.insert(timeInfo->outputBufferDacTime - soundCardTimeNow);
+
+        m_absTimeWhenPrevOutputBufferReachesDac = filteredHostTimeNow +
```
I am not sure if this works. 


Why not just follow the example? It use the same clock which is used in captureAudioSessionState() which is required:  

https://github.com/Ableton/link/blob/b6d5c597a1027f333a06459c8d6e064d603bbe7c/examples/linkaudio/AudioPlatform_Portaudio.ipp

I think we can do a better alternative when the API provides the timing: 


captureAudioSessionState() has a clock for reference and does a snapshot for "now". 
All data is extracted by passing m_absTimeWhenPrevOutputBufferReachesDac to transpose the result to the actual point of interest and not "now"

This means we actually need the time form taking "now" to the to the point when reaching the DAC. 
This varies each callback and this variation should be part of the value to actually compensate such issues.   

Summing up the DAC timing and than comparing with the wall clock does not work, we can however use the provided DAC timing to calculate the jitter as diff value and add it to the wall clock timestamp. 

In this case we need the unfiltered value to effectively compensate jitter noise. 

https://github.com/Ableton/link/blob/b6d5c597a1027f333a06459c8d6e064d603bbe7c/include/ableton/link/Controller.hpp#L239

JoergAtGithub replied:
> I am not sure if this works.

It 100% does, I spend months of testing until I reached this point. This is the essence of the whole PR!

> Why not just follow the example?

I started with the example, and it did not work very well. Even the Ableton developers wrote somewhere that they never got the PortAudio implementation working compareable to the other backends.


> We have already solved the issue when syncing the wavefoms in`updateCallbackEntryToDacTime()` This is pretty much the same task of predicting the reach of the DAC and sort out jitter noise and clock differences.

This trial-and-error approach is completely unsuitable here, as it provides neither a steady increase nor jitter-free timing information. Any errors must be averaged out by the filter ignoring individual buffer callbacks is not an option here. And this delivers an absolute time with a jitter far lower than anywhere else in Mixxx.

The implemented algorithm is tested for very long time (there is a reason, why this PR took me 4 years), and I tested many variants starting from the official Ableton examples. Believe me that this is far better than what I started from!

daschuer replied:
>> I am not sure if this works.

> It 100% does, I spend months of testing until I reached this point. This is the essence of the whole PR!

Oh sorry. This shall not come along as general critic of this PR. It was meant as inline comment for the local scope. 

I obviously do not understand the code and the reasons why it shall be like that. As a reviewer I like to understand it. This is required to verify all edge cases. Ideally it results in some nice code comments that also secures your implementation against future changes form other authors. 

So let's try to get a common understanding: 

> I started with the example, and it did not work very well. 

What was the issue? 

> Even the Ableton developers wrote somewhere that they never got the PortAudio implementation working compareable to the other backends.

Do you have a reference for this? This is probably an extremely important information for our GSoC project. 
Lets not repeat such PortAudio issue.

I know that the timing information of PortAudio is broken when we uses the ALSA loop back pulse or pipewire devices. The ALSO timing of the hardware soundcards is very stable. 
Which API do you use on windows? For my understanding native ASIO and WDM-KS also provide good timing even though PortAudio.

> Any errors must be averaged out by the filter ignoring individual buffer callbacks is not an option here.

Can you explain the reasons?

JoergAtGithub replied:
I added the requested explanations as comments!

JoergAtGithub committed:
Added good detailed comment for non obvious isPlaying method

JoergAtGithub committed:
Change destruction order in AbletonLink destructor
Destroy m_pLink before control objects

JoergAtGithub commented on src/engine/sync/abletonlink.cpp:
```diff
@@ -0,0 +1,208 @@
+}
+
+double AbletonLink::getBeatDistance() const {
+    const auto sessionState = m_pLink->captureAudioSessionState();
```
This must be captured every time again, as the world around is not stoping while the Mixxx engine is processing a buffer callback.

daschuer replied:
I think we need first find a common understanding of the issue. So please correct me if I am wrong:  

We have two independent threads firing asynchronously. The audio callback prepares the samples that are passed to the DAC ahead of time. The runtime of the Audio callback itself must not matter, if it is at least finished before the next audio callback is scheduled. This means a timing value gathered at the beginning of the callback for let's say deck 1 shall have the same timestamp in its calculation as one that is gathered later for let's say deck 4. Both produce samples that are played simultaneously and shall use the same beat distance to Link clients. 

You're right that the Link timeline doesn't stop and we want to stay responsive to network updates. However, we need a reliable relation between the audio and the link thread. A kind of freeze, process, continue.  

The core issue is this: Within a single audio callback, `getBeatDistance()` is called multiple times by different parts of the Mixxx engine. Each callback need to evaluates the session state at the freezed time, a fixed timestamp that is not effected by code runtime differences or CPU load. 

But when `getBeatDistance()` is called again moments later in the same callback, it's asking Link the same question: "What will be the beat distance in the when the buffer I prepare now is processed in the DAC" 
But if `captureAudioSessionState()` is called again between those two `getBeatDistance()` Link is advanced and we use a new session state no longer related to the first one used during the first part of the audio callback. This violates the a fundamental assumption: within one audio callback, all calculations referencing the same timestamp should be consistent and causes subtle timing glitches or phase discontinuities.

So I think we need to consider where the freeze is required and where we need a fresh timing. Here we need to consider the DAC timing. We need to compensate all CPU runtime differences and remove all time that comes from CPU processing from the calculation. We need to capture once each audio callback. 

The key is ensuring all audio calculations within one callback see a **consistent snapshot** while still allowing the system to respond to Link updates at callback boundaries.

Does that make sense?

JoergAtGithub replied:
I just tried to change this to make sessionState a memeber variable of the AbletonLink class. But it completly broke the functionality, as these member functions are also called outside of the engine callback.
The local capturing of the sessionState is clearly the correct implementation!

JoergAtGithub commented on src/util/hosttimefilter.h:
```diff
@@ -0,0 +1,86 @@
+#pragma once
```
This was what I used initially, but I decided to implement it myself, as the other implementation pulled in a lot of dependencies and did not guarantee to use the same host time clock as the other parts of Mixxx.

JoergAtGithub commented on src/engine/sync/abletonlink.cpp:
```diff
@@ -0,0 +1,208 @@
+    }
+
+    // Note, that ableton::Link::SessionState.isPlaying() is an optional Ableton
+    // Link feature (Start/Stop sync) and shouldn't be taken into account here.
```
Done!

daschuer commented on src/engine/sync/abletonlink.cpp:
```diff
@@ -0,0 +1,216 @@
+
+namespace {
+const mixxx::Logger kLogger("AbletonLink");
+constexpr mixxx::Bpm kDefaultBpm(9999.9);
```
kDefaultBpm is defined in enginesync.cpp diffrently. Is this  just a random invalid virgin value? 
mixxx::Bpm is by default kValueUndefined which is the invalid value 0.0. Can we just use this and remove the initialization with 9999.9 altogether?

JoergAtGithub replied:
Good point, I removed this constant.

JoergAtGithub committed:
Remove unnecessary kDefaultBpm. m_oldTempo now uses default Bpm initialization (kValueUndefined = 0.0)
instead of arbitrary 9999.9 value.

JoergAtGithub committed:
Improve comments to explain the determinition of the nearly-jitter free time reference.
Added comments about the weaknesses in timepoint and the latency reporting.
Explained why the official AbletonLink example for PortAudio is not working well.

daschuer commented on src/engine/enginemixer.cpp:
```diff
@@ -349,7 +358,7 @@ void EngineMixer::processChannels(std::size_t bufferSize) {
     // After local bpms are updated, trigger the rest of the post-processing
     // which ensures that all channels are updating certain values at the
-    // same point in time. This prevents sync from failing depending on
+    // same point in time.  This prevents sync from failing depending on
```
extra space

daschuer commented on src/soundio/sounddeviceportaudio.cpp:
```diff
@@ -1142,6 +1166,71 @@ void SoundDevicePortAudio::updateCallbackEntryToDacTime(
+    }
+
+    if (CmdlineArgs::Instance().getDeveloper()) {
+        qWarning() << "Pa_GetStreamTime: "
```
We must not  print in the engine thread because it is a locking operation, xcept there is a strong reason for it. I suggest to maintain this as commted qDebug output. When we need it for debugging we can uncomment compile and enjoy.

daschuer commented on src/soundio/sounddeviceportaudio.cpp:
```diff
@@ -1142,6 +1166,71 @@ void SoundDevicePortAudio::updateCallbackEntryToDacTime(
+
+    // With the cumulated buffer time an the linear regression in the host time filter we smooth
+    // and host time point for each soundcard callback - with a jitter of some 100us.
+    auto filteredHostTimeNow = m_hostTimeFilter.calcHostTime(m_cummulatedBufferTime);
```
The type can't be guessed and is important to understand the following code. 
```suggestion
    std::chrono::microseconds filteredHostTimeNow = m_hostTimeFilter.calcHostTime(m_cummulatedBufferTime);
```
It took me a while to understand. I think I got it now. 
For my understanding the output is not `filteredHostTimeNow` because the input is not "time now". The input is the jitter free time DAC time. So the output is the DAC time transposed to Host time (+ the average call delay). Somehow the Host time for the currently received buffer. If we subtract the measured `hostTime` we see the different delays (jutter) that may happen due to CPU load or other high priority task.   

Can you rename it to `bufferTimeAsHostTime` or such?

daschuer commented on src/util/hosttimefilter.h:
```diff
@@ -0,0 +1,86 @@
+
+    // Calculates the host time based on the auxiliary time using linear regression.
+    // Returns kInvalidHostTime if there are not enough points or if the calculation isn't possible.
+    std::chrono::microseconds calcHostTime(double auxiliaryTime) const {
```
insertTimePoint() and calcHostTime() are always called after each other. You may create function the does the work of both but faster than the single functions.

daschuer commented on src/soundio/sounddevicenetwork.cpp:
```diff
@@ -528,6 +538,24 @@ void SoundDeviceNetwork::updateCallbackEntryToDacTime(SINT framesPerBuffer) {
+    if (filteredHostTime != HostTimeFilter::kInvalidHostTime) {
+        m_absTimeWhenPrevOutputBufferReachesDac = filteredHostTime + outputLatency;
+    } else {
+        m_absTimeWhenPrevOutputBufferReachesDac = hostTime + outputLatency;
```
externalSyncLatencyUs is missing here. The user may also need a manual sync.

daschuer commented on src/soundio/sounddevicenetwork.cpp:
```diff
@@ -528,6 +538,24 @@ void SoundDeviceNetwork::updateCallbackEntryToDacTime(SINT framesPerBuffer) {
+    auto hostTime = std::chrono::duration_cast<std::chrono::microseconds>(
+            ClockT::now().time_since_epoch());
+
+    m_hostTimeFilter.insertTimePoint(static_cast<double>(currentTime), hostTime);
```
currentTime is already the host time with the same jitter as `hostTime`. So I assume this filter has no effect here. You need to use the calculated buffer time here like you did in Portaudio.

daschuer commented on src/soundio/sounddevicenetwork.cpp:
```diff
@@ -528,6 +538,24 @@ void SoundDeviceNetwork::updateCallbackEntryToDacTime(SINT framesPerBuffer) {
+
+    // Use HostTimeFilter class to create a smooth linear regression
+    // between absolute network time and absolute host time
+    auto hostTime = std::chrono::duration_cast<std::chrono::microseconds>(
```
We have above already m_clkRefTimer which is called at the very first (best) place. You can add a getter() to reuse m_startTime here and avoid a second ClockT::now() call.

daschuer commented on src/soundio/sounddeviceportaudio.cpp:
```diff
@@ -1142,6 +1166,71 @@ void SoundDevicePortAudio::updateCallbackEntryToDacTime(
+    // the time points.
+    m_cummulatedBufferTime += bufferSizeSec;
+    auto hostTime = std::chrono::duration_cast<std::chrono::microseconds>(
+            ClockT::now().time_since_epoch());
```
Same here, you can reuse m_clkRefTimer here

daschuer commented on src/soundio/sounddeviceportaudio.cpp:
```diff
@@ -1142,6 +1166,71 @@ void SoundDevicePortAudio::updateCallbackEntryToDacTime(
+            m_pStream); // There is a delay & jitter to timeInfo->currentTime
+
+    // The time points reported by PortAudio (timeInfo->currentTime or PaStreamCallbackTimeInfo)
+    // are not reliable (on Windows no backend reported satisfactory values - all had huge jitter).
```
On Linux ALSA and Pipewire the are reliable. So I think we can skip our own filter here. 
You can see it by the smooth waveform. Linux smooth, Windows jittery.

daschuer replied:
On the other hand the filter is fast, compared to all the other things we do in the engine. Mabe we can use it anyway ... ??

daschuer commented on src/soundio/sounddeviceportaudio.cpp:
```diff
@@ -1142,6 +1166,71 @@ void SoundDevicePortAudio::updateCallbackEntryToDacTime(
 
+    // Use HostTimeFilter class to create a smooth linear regression
+    // between absolute sound card time and absolute host time
+    PaTime soundCardTimeNow = Pa_GetStreamTime(
+            m_pStream); // There is a delay & jitter to timeInfo->currentTime
```
Use timeInfo->currentTime, it is the same, just measured at the very beginning of the callback.

daschuer commented on src/soundio/sounddeviceportaudio.cpp:
```diff
@@ -1142,6 +1166,71 @@ void SoundDevicePortAudio::updateCallbackEntryToDacTime(
+    // which the user can set in the Sound Preferences.
+    if (callbackEntrytoDacSecs > kMinReasonableAudioLatencySecs &&
+            timeSinceLastCbSecs < bufferSizeSec * 2) {
+        m_meanOutputLatency.insert(timeInfo->outputBufferDacTime - soundCardTimeNow);
```
```suggestion
        m_meanOutputLatency.insert(callbackEntrytoDacSecs);
```

daschuer replied:
Scratch that it. It was wrong. Pa_GetStreamTime(m_pStream); return now() from the soundcard clock.

daschuer commented on src/soundio/sounddeviceportaudio.cpp:
```diff
@@ -1142,6 +1166,71 @@ void SoundDevicePortAudio::updateCallbackEntryToDacTime(
+            timeSinceLastCbSecs < bufferSizeSec * 2) {
+        m_meanOutputLatency.insert(timeInfo->outputBufferDacTime - soundCardTimeNow);
+
+        m_absTimeWhenPrevOutputBufferReachesDac = filteredHostTimeNow +
```
The sanity check for wrong API data below is not considered here. I think it was required of DirectSound on windows.

daschuer reviewed (changes requested):
Some more comments. 

The filter for measuring the current buffer time works nice. 

The surrounding code is redundant with the already existing timing code for waveform anti jitter. We need to join it to make the waveforms smooth on windows as well. 
On Linux we can skip the filter because ALSA and Pipewire has its own reliable timing 

I guess the the mentioned issues with Portaudio And ableton Link are widows only issues.

daschuer commented:
This is Pipewire 0.3.48 via Pulse API (Bad timing) 
* timeInfo->outputBufferDacTime - soundCardTimeNow 
* m_meanOutputLatency.mean()
<img width="1351" height="436" alt="image" src="https://github.com/user-attachments/assets/7e0dc13a-37d4-4f13-b8fb-a7717a8177f9" />

This is the corresponding: 
* hostTime.count() - filteredHostTimeNow.count() 
<img width="1347" height="409" alt="image" src="https://github.com/user-attachments/assets/a688aafd-6f4b-47e3-90eb-d4b3886c9492" />

daschuer commented:
This is the same with ALSA:

<img width="1360" height="851" alt="image" src="https://github.com/user-attachments/assets/165eb139-c258-4aa3-8f53-0c5e475a9407" />

daschuer commented:
Takeaways from the graphs: 
ALSA:
* Direct ALSA delivers good timing 
* In the Direct ALSA timing, we probably see the scheduler quantization of 1 ms. 
* The reporter value matches the host filtered value.
 
ALSA Pulse Pipewire ALSA stack
* The delivered timing is bumpy
* Some in instances in the stack are trying to compensate a drift
* In the host measured timer we see the same "desired" 1 ms jitter,  but also a big slope trying to compensate something. Maybe a missed frame or such. 

We can also see that the time constants from both filters do not match. We need to fix the issue that we subtract two values filters with different time constances. 

If we use the filtered value + the reported time, we overcompensate the jitter noise.

DatanoiseTV commented:
Thanks for digging into the PipeWire/Pulse vs. ALSA timing — the two-filter time-constant mismatch is a nice catch.

I took a small stab at exactly that part on a branch: [`feat/ableton-link`](https://github.com/DatanoiseTV/mixxx/tree/feat/ableton-link) ([0bf8ffd](https://github.com/DatanoiseTV/mixxx/commit/0bf8ffd652e08d31b96c6cf808e1a99a2d7d48fb)). It derives the host-time-at-DAC from the single host-time regression, evaluated at `cumulatedBufferTime + outputLatency`, and matches the latency-filter window to the regression window — so the "now" filter and the latency filter no longer fight each other. On CoreAudio/ASIO/direct-ALSA it's numerically near-identical, so it shouldn't regress the cases that already work well.

The honest caveat: I could only confirm it builds and behaves on CoreAudio. I don't have a PipeWire/Pulse-over-ALSA rig with a Link peer and the scope-trace harness you used above, so I can't tell whether it actually helps the bumpy case. If someone with that setup could give it a spin, that'd be hugely helpful. Happy to turn it into a proper PR against this branch if it pans out.
